A33-Fault-Injection.md

```diff
@@ -0,0 +1,253 @@
+
+We still accept fault injection settings with only HTTP status code. **The ConfigSelector should translate the HTTP status code into gRPC status code** before sending the config to the fault injection enforcer.
+
+
+### No multiple fault injection filter
+
+Multiple fault injection is allowed in Envoy. Fault injection filters are normal HTTP filters in Envoy. All HTTP filters are executed consequently. Confirmed with experiment that 2 fault injection filters work fine. Also, if Envoy specified any Runtime fault injection parameter, it overrides all HTTP fault injection filters and only one fault injection will be working.
+
+In TD and Istio, there could only be one fault injection setting in routing configuration. Each RPC can only have 1 matched routing configuration. The fault injection policy in the routing configuration will be executed. For simplicity, **this doc proposes to only allow one fault injection filter**.
+
+
+### Evaluate Possibility Fraction
+
+According to the [Envoy implementation](https://source.corp.google.com/piper///depot/google3/third_party/envoy/src/source/extensions/filters/http/fault/fault_filter.cc;l=223), each feature of fault injection is checked with a different random number. Compared to using a single random number for all features, this affects the probability of an RPC being affected by fault injection. For example, the fault injection has 20% chance to delay and 5% chance to abort. Under the single random number solution, only 20% RPC will be affected. But if two features are independent, 24% RPC will be affected. So, **this doc suggests to make each feature independent by using a different random number each time.**
+
+
+## Alternatives
+
+### Alternative 1: Fault Injection Config via ProtoBuf Metadata
+
+After ConfigSeletor calculated the final fault injection config, it can encode the config proto and set it as metadata for that RPC. This approach is straightforward and allows more cohesive logic in the fault injection enforcer. On the other hand, it requires additional serialization and deserialization of proto messages, and it can reduce debuggability.
+
+```bash
+grpc-fault-injection-policy-bin=<...>  # The encoded FI config proto message
+```
+
+However, setting additional metadata may leak implementation details onto wire, which is undesired.
+
+
+### Alternative 2: Fault Injection Config via Multiple Metadata
```

Comment:
We basically are already going to support this due to supporting configuring FI via headers.

Reply:
This section is removed. It suggested an implementation alternative to always translate fault injection policy into gRPC metadata. It's similar to our config-FI via headers.